extensions/amp-viewer-integration/0.1/messaging.js

```diff
@@ -42,6 +42,38 @@ const MessageType = {
  */
 export let Message;
 
+export class WindowPortEmulator {
+  /**
+   * @param {!Window} win
+   * @param {string} origin
+   */
+  constructor(win, origin) {
+    /** @const {!Window} */
+    this.win = win;
+    /** @private {string} */
+    this.origin_ = dev().assertString(origin);
+  }
+  /**
+   * @param {string} eventType
+   * @param {function(!Event):undefined} handler
+   */
+  addEventListener(eventType, handler) {
+    listen(this.win, 'message', e => {
+      if (e.origin == this.origin_ &&
+          e.source == this.win.parent && e.data.app == APP) {
```

Comment:
I mentioned before, and we could probably delay this a bit. But you'd later need to pass `source` into this class again, since it's not always true that `source == win.parent`. But again, it can be delayed.